### DISCUSSION TITLE
Wrapped more 2d array message filter

### DISCUSSION THREAD

**Discussion by zeyus** [FLAMEGPU/FLAMEGPU2#894] [Q&A] [answered]
in order to limit memory usage, I'm trying to make a moore sequence function, basically to map the `message_in`  sequence into an array of `flamegpu::id_t`.

```cuda


FLAMEGPU_AGENT_FUNCTION(get_game_list, flamegpu::MessageArray2D, flamegpu::MessageNone) {
    const unsigned int my_x = FLAMEGPU->getVariable<unsigned int>("x_a");
    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int>("y_a");
    const unsigned int my_id = FLAMEGPU->getID();

    // iterate over all cells in the neighbourhood
    // this also wraps across env boundaries.
    unsigned int num_neighbours = 0;
    unsigned int neighbour_id = 0;
    unsigned int num_responders = 0;
    for (auto &message : FLAMEGPU->message_in.wrap(my_x, my_y, 1)) {
        flamegpu::id_t competitor_id = message.getVariable<flamegpu::id_t>("id");
        // challenge competitor if competitor id is higher
        if (competitor_id > my_id) {
            // valid neighbour
            ++num_neighbours;
            // we will challenge them
            ++num_responders;
            FLAMEGPU->setVariable<flamegpu::id_t, 8>("game_list", neighbour_id, competitor_id);
        } else {
            if (competitor_id != flamegpu::ID_NOT_SET) {
              // valid neighbour, but they are our challenger
              ++num_neighbours;
            }
            FLAMEGPU->setVariable<flamegpu::id_t, 8>("game_list", neighbour_id, flamegpu::ID_NOT_SET);
        }
        ++neighbour_id;
    }
    // If there are no neighbours, it's time to move, otherwise let's play a game.
    if (num_neighbours == 0) {
        float my_energy = FLAMEGPU->getVariable<float>("energy");
        float travel_cost = FLAMEGPU->environment.getProperty<float>("travel_cost");
        // try and deduct travel cost, die if below zero, this will prevent
        // unnecessary movement requests
        my_energy -= travel_cost;
        if (my_energy <= 0.0) {
            return flamegpu::DEAD;
        }
        FLAMEGPU->setVariable<float>("energy", my_energy);
        // we have to move
        FLAMEGPU->setVariable<unsigned int>("agent_status", 32);
    } else {
        // we have to play a game
        FLAMEGPU->setVariable<unsigned int>("agent_status", 8);
    }
    FLAMEGPU->setVariable<unsigned int>("challengers", num_neighbours - num_responders);
    FLAMEGPU->setVariable<unsigned int>("responders", num_responders);
    return flamegpu::ALIVE;
}



FLAMEGPU_HOST_DEVICE_FUNCTION void pos_from_moore_seq(const unsigned int x, const unsigned int y, unsigned int &move_index, unsigned int &new_x, unsigned int &new_y, const unsigned int env_max) {
    // uniform int represents the direction to move,
    // e.g. for radius 1, 0 = northwest, 1 = west, 2 = southwest, 3 = north
    // (4 = no movement), 5 = south, 6 = northeast, 7 = east, 8 = southeast
    if (move_index >= 4) {
      ++move_index;
    } else if (move_index > 9) {
      // wrap around the space, e.g. with radius 1, if move_index is 10, then move_index = 1.
      move_index = move_index % 9;
    }
    // Convert to x,y offsets
    const int new_x_offset = move_index / 3 - 1;
    const int new_y_offset = move_index % 3 - 1;
    // const int new_x_offset = move_index % 3 - 1;
    // const int new_y_offset = move_index / 3 - 1;

    // set location to new x,y and wrap around env boundaries
    new_x = (x + new_x_offset) % env_max;
    new_y = (y + new_y_offset) % env_max;
}


FLAMEGPU_AGENT_FUNCTION(play_challenge, flamegpu::MessageNone, flamegpu::MessageArray2D) {
    unsigned int game_sequence = FLAMEGPU->getVariable<unsigned int>("game_sequence");
    flamegpu::id_t opponent = FLAMEGPU->getVariable<flamegpu::id_t, 8>("game_list", game_sequence);

    if (opponent == flamegpu::ID_NOT_SET) {
      // no opponent in this sequence, move along
      ++game_sequence;
      FLAMEGPU->setVariable<unsigned int>("game_sequence", game_sequence);
      return flamegpu::ALIVE;
    }

    // opponent found, play a game
    const unsigned int my_x = FLAMEGPU->getVariable<unsigned int>("x_a");
    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int>("y_a");

    // opponent location
    unsigned int opponent_x;
    unsigned int opponent_y;

    // this is dumb, but it's passed as a reference
    unsigned int local_game_seq = game_sequence;
    pos_from_moore_seq(my_x, my_y, local_game_seq, opponent_x, opponent_y, 32);

    // set message to opponent
    FLAMEGPU->message_out.setVariable<flamegpu::id_t>("id", FLAMEGPU->getID());
    // no other agent should have the same opponent this round
    // so no issue with multiple messages out
    FLAMEGPU->message_out.setIndex(opponent_x, opponent_y);

    const unsigned int num_challengers = FLAMEGPU->getVariable<unsigned int>("challengers");
    unsigned int num_responders = FLAMEGPU->getVariable<unsigned int>("responders");
    // decrement responders
    --num_responders;


    // update agent
    if (num_responders + num_challengers <= 0) {
        FLAMEGPU->setVariable<unsigned int>("agent_status", 16);
    } else {
      // we don't need to waste a variable set op
      ++game_sequence;
      FLAMEGPU->setVariable<unsigned int>("game_sequence", game_sequence);
    }
    FLAMEGPU->setVariable<unsigned int>("responders", num_responders);

    return flamegpu::ALIVE;
}

```

For my movement submodel, it seems to work perfectly, and agents move (starting with a random `moore index`) until they have another agent within the moore neighbourhood.

The problem I'm getting is with the previous layer (in a different submodel), which is the code outlined above. All the hardcoded `int` values come from python variables set outside the cuda code (I'm trying to do that with any variables that won't change over a bunch of simulations).

When the agent `play_challenge` function is included, FGPU throws this error: `Array messagelist contains 12 messages at index 0!`

Again, this must be something silly I've done (or another misunderstanding, or both), but as I understand it, if the agents have a sequence of runs in the submodel, their variable `game_sequence` will increment each time, and each increment should map to either `flamegpu::ID_NOT_SET`, or `flamegpu::id_t` with the agent ID that is currently in that space.

And if the execution of the game play is sequential, each agent should play either nobody, or another agent in the same direction as all the other players for that step, which should make multiple messages to the same index impossible, yet, the fact that it's happening points to some major flaw in my implementation haha.

I looked at the code (https://docs.flamegpu.com/api/program_listing_file_include_flamegpu_runtime_messaging_MessageArray2D_MessageArray2DDevice.cuh.html), and as I understood it, the moore wrap works in the following way for radius of 1:

| 0 | 3 | 5 |
:----:|:----:|:----:
| 1 | x | 6 |
| 2 | 4 | 7 |

Full code here, if it's necessary, but it's...not exactly pretty: https://github.com/zeyus/SoCultABM/blob/OL_PD/src/pd/model.py

If I can get my head around this and get this method working, I'll retroactively apply it to the movement, because I think it should more efficient, even though it might require a max of `(message_radius + 2)^2 - 1` "iterations" per step (as long as collisions aren't being introduced, and the number will decrease  each time as agents move in the direction of the iteration) and will use way less memory than a `env_max x env_max` macro array property.

Maybe another approach might be to add another message layer and have a function input a message array, and check if there is any move attempt, and broadcast a new location...given the current configuration, agents only move if they have no neighbours, which makes it impossible for every space around them to be taken after agents move, there can only be a maximum of 4 agents trying to move to a particular grid location.

Movement only with the game main model layers commented out (movement uses the same `pos_from_moore_seq` function definition):
![move_b2](https://user-images.githubusercontent.com/75656/182975143-f873c56a-e9d8-4d85-b94e-b61a389a6ed8.gif)

**Comment by Robadob** [comment, marked as answer]
> When the agent play_challenge function is included, FGPU throws this error: Array messagelist contains 12 messages at index 0!

I expect this refers to 12 agents calling `FLAMEGPU->message_out.setIndex(opponent_x, opponent_y);` with the arguments `(0,0)`. It's thrown from a [common method](https://github.com/FLAMEGPU/FLAMEGPU2/blob/ca209d004f47de5a39a6b4fdb7404cd1cf3853fe/src/flamegpu/gpu/CUDAScatter.cu#L618) used by all 3 forms of array message, so the dimensions aren't available (it sees them all as 1D).

It probably also checks bins of the array list in order, so highly likely you have collisions in other elements of the arraylist too.

As you've said in your other comment, you probably want agents to output a message at their own location, rather than at their desired location (as there might be a conflict there). If you want to support multiple messages at a location, you will need to switch to `MessageBucket`, but that will require you to convert 2D coordinates to 1D manually.

**Reply by zeyus** [reply]
Right, which points to a major bug (in my code), because `opponent_x`, `opponent_y` could be 0,0 but only one should ever be 0,0 at a time. If all agent's start by playing a responder northwest of them, then all west, etc, there couldn't be an overlap (i.e. there is only one agent directly southeast or east), so either the values are getting set wrong, or there's a condition in my function that sets the variables to 0 (though my python mock didn't seem to show that behaviour).

I'm rewriting it to output to their own location now, as multiple messages at the same location will still have to deal with play order and agents that die.

Again, appreciate the help, I feel like I'm starting to get a more complete understanding of FGPU2 and getting to be a bit more comfortable with (CU)CPP which is a nice bonus :)

**Reply by zeyus** [reply]
Ok, so, after refactoring the code, and only outputting to the agent's current location, I'm still getting the same error...

Is there any case where the `MessageArray2D` somehow persists?

Currently I have
Main model:
 - Layer 1: broadcast current location
 - Layer 2: receive broadcast and look for neighbours
 - Layer 3: game submodel
 - - Layer 1: Challenge message (at current agent location) for a single responder
 - - - Function cond: num_responders > 0
 - - Layer 2: respond to neighbouring challenges (should only be one, but that's irrelevant to array index)
 - - - Function cond: num_challengers > 0
 - - Exit cond: no agents playing
 - Layer 4: movement submodel
 - - Layer 1: request movement
 - - - Function Condition: Agent status movement unresolved
 - - Layer 2: respond to movement
 - - - Function condition: agent status moving
 - - exit cond: no agent's moving

Writing this out, is it possible that the layer executes WHILE any agents matching the function condition still exist /before/ moving to the next layer?

I'll test it by adding an extra condition (responders > 0 && some_status)

**Reply by Robadob** [reply]
> https://github.com/zeyus/SoCultABM/blob/OL_PD/src/pd/model.py

I guess this repo is private? 404's for me

**Reply by zeyus** [reply]
> > https://github.com/zeyus/SoCultABM/blob/OL_PD/src/pd/model.py
> 
> I guess this repo is private? 404's for me

Oops, I forgot I had it private, sorry. It's public now...as it should be!

**Reply by Robadob** [reply]
Will try and get time to give it a play later today.

**Reply by zeyus** [reply]
Amazing, I'll be working on a solution/workaround/comprehension, and I'll post here as soon as I get it.

**Reply by zeyus** [reply]
> I'll test it by adding an extra condition (responders > 0 && some_status)

I've tried a couple of ways, and currently the agent status changes no matter what (so it doesn't meet the function condition), and then after the second layer, the same happens so they either have completely finished all their games, or they will have a status that allows them to meet the layer 1 condition.

I'm more convinced now this is one of those 'been staring at the screen too long' issues where I've assigned some function variable, message variable, or layer variable in the python code wrong, but I've tried going through it line by line and I can't see any accidental variable redefinitions or setting the wrong function condition etc.

The only other thing would be if the message array persists, but the movement submodel which follows a similar process, can potentially output a message to the agents current cell over several layer1 iterations if the first move is rejected, and the movement model  doesn't have those collisions in my experimentation so far (even with large numbers of agents to increase the chance that two will request the same space)

**Reply by Robadob** [reply]
Aha, just started having a quick look at it. You probably haven't enabled optional message output. I presume agents that don't output would default to index 0.

Again, mandatory message output is default. Because processing optional output incurs a cost.

~~And this doesn't appear to be mentioned on [the most relevant page of the userguide](https://docs.flamegpu.com/guide/defining-messages-communication/index.html).~~

Ah, it's explained [here](https://docs.flamegpu.com/guide/agent-functions/agent-communication.html#sending-messages).

**Reply by zeyus** [reply]
> Aha, just started having a quick look at it. You probably haven't enabled optional message output. I presume agents that don't output would default to index 0.

🤦‍♀️omg, that's fixed it completely.

and the second I saw your response here, I remember reading it in the docs (I've read the docs so many times now), but it totally slipped my mind.

Makes perfect sense, I'll see how the model performs as-is, and get it working, and if it seems like it's slow I can try changing to a message bucket, and just have a bin for "no play/ignore"

The amount of time I spent trying out 100 different approaches, and the answer was there all along. :D

**Reply by Robadob** [reply]
After that's fixed, your model crashes because in `exit_play_fn` you are calling `FLAMEGPU.agent("prisoner")` which attempts to get the `HostAPI` object for the  `default` prisoner state, where you have custom states. (Each agent state is stored in memory separately, so they have distinct HostAPI objs).

**Reply by zeyus** [reply]
Yeah, I noticed that too. I'll probably undo adding states entirely, (as a quick workaround I added the "default" state explicitly). I was looking at that as an alternative to prevent layer / submodel execution, but to do that I would need to be able to make one function have agents end in either "playing" or "moving" states, and I didn't see anything in the DeviceAPI that lets you set the agent state.

In order to implement it (if there's even value in doing so) I guess I'd need two separate functions, each with a condition that will only match a subset of agents, and then set the end state for those agent functions respectively.

**Reply by zeyus** [reply]
One addendum, after spending time with FGPU2 and getting more familiar with it, I've decided that I would have been better off writing the entire model in CUDA / C++, but it seemed like it would have been out of reach (and it probably would have stopped me), I think once the exam's submitted, I might try and rewrite it and skip the python.

**Reply by Robadob** [reply]
Agents are transferred between states with agent function conditions.

https://docs.flamegpu.com/guide/agent-functions/conditional-behaviours.html

You specify a condition, only agents that pass it execute the function and transfer from the specified input-output state.

The original FLAMEGPU 1 (and it's predecessor CPU modelling framework FLAME) were designed around the idea of state machines, so it follows this flow. I don't think I've ever needed to employ states in any models I've worked with.

@AshurstJ is using states for a model he's developing as part of his PhD.

**Reply by Robadob** [reply]
>  and if it seems like it's slow

It shouldn't be 'slow', it's just slower than not doing it in the first place. In general we err on the side of avoid redundant code wherever possible.

> In order to implement it (if there's even value in doing so) I guess I'd need two separate functions, each with a condition that will only match a subset of agents, and then set the end state for those agent functions respectively.

That seems logical to me, but I'm really not upto speed on states. I set up all that stuff years ago and haven't looked at it since.

> I've decided that I would have been better off writing the entire model in CUDA / C++, 

RTC (RunTimeCompiled, e.g. Python) models actually run faster than C++. Run time compilation allows us to do some extra post-processing to make variable accesses even cheaper.

There is very little you can't do in the Python API which you can in the C++. Bespoke reductions are the only thing which springs to mind.

You might find it useful to split your agent function sources out into separate `.c` or `.cu` files though. This will give you some degree of syntax highlighting, then just use [`newRTCFunctionFile()`](https://github.com/zeyus/SoCultABM/tree/OL_PD/src/pd) instead. Likewise, I prefer to keep submodels in separate files to try and keep things distinct (This does work in Python, just make sure host functions are disowned).

**Reply by zeyus** [reply]
> It shouldn't be 'slow', it's just slower than not doing it in the first place. In general we err on the side of avoid redundant code wherever possible.

Fair enough, I'm sure it'll be performant enough, and for the purpose of this exam it doesn't really matter, we can cap the population at whatever number explodes my GPU (although a friend did offer me some time on his 3080).

> RTC (RunTimeCompiled, e.g. Python) models actually run faster than C++.

Huh, interesting, that seems initially counterintuitive, but it makes sense.

> There is very little you can't do in the Python API which you can in the C++.

No it's more the mix of languages and the inlined agent functions, but as you say

> You might find it useful to split your agent function sources out into separate .c or .cu files

Which I was initially doing, and I might change it back, but I feel like there's some benefit (depending on how many different parameters we will end up tweaking for this model) in allowing variables to be set outside the code and just replacing them inside the CUDA functions. Again, the difference is probably minimal for this kind of thing, but my thought process was avoiding unnecessary calls inside the functions to retrieve environment/macro properties, while still allowing them to change globally. Of course I could do some weird templating thing where if a `.tmpl.cu` exists, read it in, replace the variables from python and save it as `.cu` but that seems kind of gross haha.

I haven't investigated much, but if it's possible to include some "globally" (i.e. within all agent functions) available functions, or declare a new class with some static properties, or even just consts that could be referenced inside the agent functions, the same result would be achieved. I stopped going down that rabbit hole when I tried to declare an array from `std` lib and realised that wasn't going to work.

**Reply by Robadob** [reply]
Modern CUDA does support `std::array`, but runtime compiled CUDA doesn't (yet?). I think I have a development branch somewhere that I added support for `std::array` but it never got merged as RTC didn't like it. C style arrays should work though.

**Reply by Robadob** [reply]
> I haven't investigated much, but if it's possible to include some "globally" (i.e. within all agent functions) available functions, or declare a new class with some static properties, or even just consts that could be referenced inside the agent functions, the same result would be achieved. I stopped going down that rabbit hole when I tried to declare an array from std lib and realised that wasn't going to work.

RTC Agent functions should support includes, so you could have a common header with alot of defines e.g. `#define FOO 12`. Modern C++ constexpr may also work, but less sure on that.

[Details Here](https://github.com/FLAMEGPU/FLAMEGPU2/pull/801), documenting this is an [open issue](https://github.com/FLAMEGPU/FLAMEGPU2-docs/issues/80)

**Reply by zeyus** [reply]
> RTC Agent functions should support includes, so you could have a common header with alot of defines e.g. `#define FOO 12`.

That's awesome, then when I get the model running, or when I get sick from reading brown text, I'll refactor them out, and even if the variables should stay in python, generating a single `.cuh`/`.h` with the required defines from python would be trivial :)
